### ISSUE TITLE
Suggestion: immutable list

### ISSUE THREAD

**Issue by MeeniMc** [gbl/EasierCrafting#33]
# TL:DR

* **Goal: Recipes should not move all around the screen** without first closing-reopening the inventory. Even when crafting recipes exhaust or add ingredients to  the inventory. 

* **Proposition**: when the inventory is opened, the **list of 'active' recipes is immutable** (a recipe that exhaust its ingredients just become gray/crossed and clicking it has no effect), and recipes that become active get **appended at the end of the list.**

# Details of the problem

Recipes can move up and down in the list when the content of the inventory change due to crafting some item. The required components may become unavailable (clicking the recipe (or another recipe) consumed the ingredients), or new recipes may become active (clicking a recipe produced an ingredient to a new recipe). This will reshuffle the list of recipes and cause clicking on the wrong recipe and craft the wrong item. 

# Shortcomings of how the problem is currently addressed

I understand that you are aware of this problem, given that you have various workaround strategies in the mod. 
* The shift-click strategy (produce everything that is possible) removes most of the pain for this problem, and it is great when it is what is needed. It does not help when say I have enough to produce 14 hoppers, but what I actually want is 7 chests and 7 hoppers. I have to click 7 times, and at every click, the recipe list may change under my cursor, and I might end-up crafting something useless. 
* The 5s cooldown (default) helps, but I may still need to click for more than 5s on a given recipe, if I need to produce say 40 or more (but not 64, see above), in which case the recipe will still change while I am clicking.
* The 5s cooldown is actually annoying when chaining recipes (say, craft a bunch of chests, then craft a bunch of hoppers). But reducing it to a non-annoying value is dangerous. 
* One may still use the vanilla craftbook to tackle these cases, but having to do two different ways of crafting is not ideal, and any miscalculation can cause crafting wrong items. 

# Details of the proposition for solving the problem

1. When the inventory is opened, the list of recipe is produced the same it is currently.
2. If crafting causes an ingredient to a recipe to become depleted, the recipe **remains in the list, but is gray/crossed** to indicate that it cannot be further crafted with the current inventory. Clicking the recipe has no effect. 
3. If crafting causes new recipes to become active (due to new ingredients appearing), they are appended (instantaneously) to the list. **Appended, so that the existing portion of the list is not mutated**. 

# Limitations of the proposed approach

1. User may have to close-reopen the inventory to purge "inactive" recipes for which ingredients are no more present. This is IMO a desired behavior.
2. The 'append' behavior for newly active recipe may cause the list to not abide by the normal sort order (categories, alphabetic, or whatever). This is displeasing, but this is less displeasing than the recipes moving around, and again, the user may close-reopen the inventory to restore sorting order. 

Again, very good mod, thanks for reading me.

**Comment by gbl** [comment]
Thank you for writing such a detailed suggestion!

However, I want to challenge the premise that "I may still need to click for more than 5s on a given recipe" - as long as you keep clicking, the timer gets reset. So If you need 40 seconds to click the hopper 40 times, the recipe list won't change until after 45 seconds. Which means, you can safely set the timer to 1 or 2 seconds, still click as often as you want (most people need less than a second for a click), and get an update much sooner.

I don't really like adding items to the existing list, because it makes things harder to find, and kills the idea of categories.

What do you think of this idea: whenever the list changes, after the existing timeout, the list becomes dark, fades in for about 0.5 seconds, and doesn't accept clicks until fully faded in? That's short enough to not be annoying, as people will need to find the item in the new list anyway, but long enough to prevent accidential crafts.

**Comment by MeeniMc** [comment]
You are right in general, as long as you click, indeed, the list update is deferred (good!). 

However I was trying to hit corner cases a bit today. Here's what I found. 

### Incomplete exhaustion of ingredients use case

If you keep clicking on a recipe when its ingredients get exhausted, obviously, no new items are produced. There are two possible cases for this:

  1. The recipe ingredients are completely exhausted (say you are producing diamonds from blocks, since the only ingredient is ONE block, it always exhaust 'all' ingredients at the same time). In this case, the behavior is as you describe: the timeout per click gets reset and if you spam click, no new updates to the list are made until you stop clicking. All is good in this case. 

  2. The recipe ingredients are incompletely exhausted (say you are producing blocks from a non-9-multiple number of diamond, or the recipe uses a variety of ingredients, and you run out of one of them only). In this case, the timeout per click is NOT reset. If you spam click, after 5s (default), the list will update itself anyway. Of note, in this case, the 'crafting area' remains cluttered with an incomplete recipe. A sort of beneficial side effect is to protect from producing unwanted items because no recipe works, the unwanted click has no effect. A less beneficial side effect is that even intentional clicks (on another recipe for which ingredients are available) have no effect in that state (one may manually remove the items from the crafting area, then resume using the mod as intended). 

I believe that if case 2 is fixed, the current approach of not updating while clicking is enough in itself and requires no further changes. 

### new recipes available use case

I did not find abnormal behavior on that side of the business. But here is the actual scenario that led me to misclick on the wrong item because of item updates. I set the timeout to 1s here.

1. Create some iron ingots from blocks
2. look at the list to find the next item to craft (planks), takes about 1s to find them
3. click on planks, but it's been a second without clicking, hence the list updates to show me that iron blocks are now available, I create an iron block. 

I agree that not respecting ordering is not ideal. I think what you propose could mitigate greatly this particular use case. Maybe an alternative, or complementary thing would be to indicate the ;refresh timeout' on the cursor (like there is a small progress bar when you attack on the minecraft cursor). That way, the cursor would itself indicate that the item you are about to click is about to change soon. That may be a sufficient addition to not require the 'dead interface for .5s' you propose.

**Comment by MeeniMc** [comment]
I'd like to revisit this proposal with the idea of CLICK-TO-REFRESH. I believe that this idea removes the problem of messing-up the categories, and the risk of clicking the wrong recipe by mistake.


1. When a recipe exhaust its recipients, it becomes gray/crossed to mark it as currently 'inactive', but it is not removed from the list.
2. When new recipients are created from other crafting actions (e.g., merging nuggets creates bars, which could turn into blocks, the list is not updated (at all), but a new indicator/button "REFRESH" appears on the side of the crafting list 
3. clicking the "REFRESH" button refreshes the list (that is, remove currently inactive recipes, adds all new possible recipes, just like the current timer-based thing do). 
4. (optional) auto-refresh and click-to-refresh (this idea) can be used at the same time. The REFRESH button is always present, and shows a countdown "Refresh in x seconds", clicking forces the refresh immediately (thus there is no wait-time when setting the auto-refresh to a large value). 
5. (optional) right clicking in an empty space is the same as clicking the refresh button. 
6. (optional) new recipes since the last "refresh" are highlighted with some shiny color/particles

**Comment by gbl** [comment]
I just implemented a "easy" version of this - whenever the list gets updated, the recipes fade out for a split second (the time for them to do this is configurable). Which means you won't accidentially craft new recipes any more after an update. Also, I set the default time to do the update to 2 seconds, so it's not worth while to close and reopen any more.
